### ISSUE TITLE
Rule Change: allow expression type annotation in `func-style`

### ISSUE THREAD

**Issue by remcohaszing** [eslint/eslint#19719]
### What rule do you want to change?

`func-style`

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
type Fn = () => undefined

/* eslint func-style: ["error", "declaration"] */

const expression: Fn = function () {}

const arrow: Fn = () => {}


[playground](https://typescript-eslint.io/play/#code=C4TwDgpgBAYgdlAvFAFASiQPigVzgEwgDMBLOCfAKEoHoAqCAZwBszgoi8BjAWkdGYQAXFADaAIggAnKQHsp4gDRRxhLswCGUjcBKy44gLp0a1Lvv5QIADzBSmjPXBHwkHbrv2oMAbygBfMwt2LTkAdxcEZHQsKD9AoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)
```

### What does the rule currently do for this code?

It reports both `expression` and `arrow` should be declared as a function declaration.

### What will the rule do after it's changed?

I suggest to add a new option `allowTypeAnnotation`. If the function style is set to `declaration`, and `allowTypeAnnotation` is true, then the rule will not report expression-style functions if the variable declarator has a type annotation.

### Participation

- [x] I am willing to submit a pull request to implement this change.

### Additional comments

I reported this a while back as https://github.com/typescript-eslint/typescript-eslint/issues/8488. But I understand ESLint core rules now also accept some TypeScript specific options.

**Comment by nzakas** [comment]
Thanks for the suggestion. The [`func-style`](https://eslint.org/docs/latest/rules/func-style) rule is currently frozen, meaning we aren't making any further changes to it and we'd need the @eslint/eslint-tsc  to override that. I believe the intent here is to make `func-style` TypeScript-aware, which seems like it goes along with our ongoing work to make ESLint as a whole more TypeScript-aware. As a result, I'm :+1: to overriding the frozen status of this rule for this particular purpose.

**Comment by JoshuaKGoldberg** [comment]
Related issues over on typescript-eslint:

* https://github.com/typescript-eslint/typescript-eslint/issues/1236
* https://github.com/typescript-eslint/typescript-eslint/issues/8488: as noted, the equivalent feature request
* https://github.com/typescript-eslint/typescript-eslint/issues/10127: also mentions https://github.com/eslint/eslint/issues/18842#issuecomment-2400677082

Would an `allowOverloads` option as proposed in 1236 be in scope for making the rule TypeScript-syntax-aware?

**Comment by nzakas** [comment]
@JoshuaKGoldberg can you summarize what you envision for `allowOverloads`?

**Comment by JoshuaKGoldberg** [comment]
I'm thinking an boolean off-by-default option that, when enabled, exempts function declarations with overloads from checking.

```ts
/* eslint func-style: ["error", { "allowOverloads": true }] */

// func-style report: Expected a function expression.
function noOverloads() {}
// ~~~~~~~~~~~~~~~~~~~~~~~

// No func-style report
function hasOverloads(value: string): string
function hasOverloads(value: number): number
function hasOverloads(value: unknown) {
  return value;
}
```

**Comment by nzakas** [comment]
Okay, I think the overloads discussion should be a separate issue as it's only related to this issue because it also deals with `func-style`. @JoshuaKGoldberg please open a separate issue for that.

Still looking for feedback from @eslint/eslint-tsc on https://github.com/eslint/eslint/issues/19719#issuecomment-2872936349

**Comment by sam3k** [comment]
We've agreed to implement this change behind a new option in May 5th TSC meeting. We've also agreed that updating a frozen rule to support TypeScript syntax makes sense.

**Comment by Tanujkanti4441** [comment]
Marking as accepted. @remcohaszing would you be open to submitting a PR for this?
